src/toxcore/dht/server/hole_punching.rs

```diff
@@ -0,0 +1,363 @@
+            "127.0.0.1:22222".parse().unwrap(),
+            "127.0.0.2:33333".parse().unwrap(),
+            "127.0.0.1:44444".parse().unwrap(),
+            "127.0.0.1:55555".parse().unwrap(),
+            "127.0.0.1:55556".parse().unwrap(),
+            "127.0.0.1:55557".parse().unwrap(),
+            "127.0.0.1:55558".parse().unwrap(),
+            "127.0.0.2:55559".parse().unwrap(),
+        ];
+
+        let mut hole_punch = HolePunching::new();
+        hole_punch.is_punching_done = false;
+        thread::sleep(Duration::from_millis(1));
+
+        assert!(hole_punch.try_nat_punch(&alice, friend_pk, addrs, Duration::from_millis(1)).wait().is_ok());
+    }
+
+    #[test]
+    fn hole_punch_get_nat_ports_test() {
+        let (pk, sk) = gen_keypair();
+        let (friend_pk, _friend_sk) = gen_keypair();
+        let (tx, _rx) = mpsc::unbounded::<(DhtPacket, SocketAddr)>();
+        let alice = Server::new(tx, pk, sk);
+        let addrs = vec![
+            "127.0.0.1:11111".parse().unwrap(),
+            "127.0.0.1:22222".parse().unwrap(),
+            "127.0.0.2:33333".parse().unwrap(),
+            "127.0.0.1:44444".parse().unwrap(),
+            "127.0.0.1:55555".parse().unwrap(),
+            "127.0.0.1:55556".parse().unwrap(),
+            "127.0.0.1:55557".parse().unwrap(),
+            "127.0.0.1:55558".parse().unwrap(),
+            "127.0.0.2:55559".parse().unwrap(),
+        ];
+
+        let mut hole_punch = HolePunching::new();
+        hole_punch.is_punching_done = false;
+        thread::sleep(Duration::from_millis(1));
+
+        assert!(hole_punch.try_nat_punch(&alice, friend_pk, addrs, Duration::from_millis(1)).wait().is_ok());
+    }
+
+    #[test]
+    fn hole_punch_punch_test() {
+        let (pk, sk) = gen_keypair();
+        let (friend_pk, friend_sk) = gen_keypair();
+        let (tx, rx) = mpsc::unbounded::<(DhtPacket, SocketAddr)>();
+        let alice = Server::new(tx, pk, sk);
+        let addrs = vec![
+            "127.0.0.1:11111".parse().unwrap(),
+            "127.0.0.1:22222".parse().unwrap(),
+            "127.0.0.2:33333".parse().unwrap(),
+            "127.0.0.1:44444".parse().unwrap(),
+            "127.0.0.1:55555".parse().unwrap(),
+            "127.0.0.1:55556".parse().unwrap(),
+            "127.0.0.1:55557".parse().unwrap(),
+            "127.0.0.1:55558".parse().unwrap(),
+            "127.0.0.2:55559".parse().unwrap(),
+        ];
+
+        let mut hole_punch = HolePunching::new();
+        hole_punch.is_punching_done = false;
+        thread::sleep(Duration::from_millis(150));
+
+        hole_punch.try_nat_punch(&alice, friend_pk, addrs, Duration::from_millis(150)).wait().unwrap();
+        let (received, _rx) = rx.into_future().wait().unwrap();
+        let (packet, _addr) = received.unwrap();
+        let mut buf = [0; 512];
+        let (_, size) = packet.to_bytes((&mut buf, 0)).unwrap();
+        let (_, ping_req) = PingRequest::from_bytes(&buf[..size]).unwrap();
```

Comment:
Use `unpack!` macro instead of this hack. You can move this macro to some common place, `binray_io`, for example.